pls/rootchain_manager.go

```diff
@@ -370,6 +424,116 @@ func (rcm *RootChainManager) handleEpochPrepared(ev *rootchain.RootChainEpochPre
 	return nil
 }
 
+func (rcm *RootChainManager) handleBlockFinalzied(ev *rootchain.RootChainBlockFinalized) error {
+	rcm.lock.Lock()
+	defer rcm.lock.Unlock()
+
+	e := *ev
+
+	log.Info("RootChain block finalized", "forkNumber", e.ForkNumber, "blockNubmer", e.BlockNumber)
+
+	// TODO: check callOpts first if caller doesn't work.
+	callerOpts := &bind.CallOpts{
+		Pending: true,
+		Context: context.Background(),
+	}
+
+	block, err := rcm.rootchainContract.Blocks(callerOpts, e.ForkNumber, e.BlockNumber)
+	if err != nil {
+		return err
+	}
+
+	privKey, err := crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+	if err != nil {
+		log.Error("Failed to get operator private key")
+		return err
+	}
+
+	transactOpts := bind.NewKeyedTransactor(privKey)
+	transactOpts.GasLimit = 4000000
+
+	if block.IsRequest {
+		invalidExits := rcm.invalidExits[e.ForkNumber.Uint64()][e.BlockNumber.Uint64()]
+		for i := 0; i < len(invalidExits); i++ {
+
+			var proofs []byte
+			for j := 0; j < len(invalidExits[i].proof); j++ {
+				proof := invalidExits[i].proof[j].Bytes()
+				proofs = append(proofs, proof...)
+			}
+			// TODO: ChallengeExit receipt check
+			tx, err := rcm.rootchainContract.ChallengeExit(transactOpts, e.ForkNumber, e.BlockNumber, big.NewInt(invalidExits[i].index), invalidExits[i].receipt.GetRlp(), proofs)
+			if err != nil {
+				log.Warn("Failed to submit challengeExit", "error", err)
+			}
+
+			log.Info("challengeExit is submitted", "exit request number", invalidExits[i].index, "hash", tx.Hash().Hex())
+		}
+	}
+
+	return nil
+}
+
+func (rcm *RootChainManager) runDetector() {
+	events := rcm.eventMux.Subscribe(core.NewMinedBlockEvent{})
+	defer events.Unsubscribe()
+
+	caller, err := rootchain.NewRootChainCaller(rcm.config.RootChainContract, rcm.backend)
+	if err != nil {
+		log.Warn("failed to make new root chain caller", "error", err)
+		return
+	}
+
+	// TODO: check callOpts first if caller doesn't work.
+	callerOpts := &bind.CallOpts{
+		Pending: false,
+		Context: context.Background(),
+	}
+
+	for {
+		select {
+		case ev := <-events.Chan():
+			rcm.lock.Lock()
+			if rcm.env.IsRequest {
+				var invalidExitsList invalidExits
+
+				forkNumber, err := caller.CurrentFork(callerOpts)
+				if err != nil {
+					log.Warn("failed to get current fork number", "error", err)
+				}
+
+				if rcm.invalidExits[forkNumber.Uint64()] == nil {
+					rcm.invalidExits[forkNumber.Uint64()] = make(map[uint64]invalidExits)
+				}
+
+				blockInfo := ev.Data.(core.NewMinedBlockEvent)
+				blockNumber := blockInfo.Block.Number()
+				receipts := rcm.blockchain.GetReceiptsByHash(blockInfo.Block.Hash())
+
+				for i := 0; i < len(receipts); i++ {
+					if receipts[i].Status == types.ReceiptStatusFailed {
```

Comment:
We have to check if `receipts[i]` is enter request or exit request, because prior undo request let the posterior enter request be reverted. `tx.data` is required to do this. 

We can just leave this as a `TODO`.